### ISSUE TITLE
Improve efficiency of `is_empty`

### ISSUE THREAD

**Issue by overlookmotel** [ParkMyCar/compact_str#354]
`is_empty` checks both inline length and heap length:

https://github.com/ParkMyCar/compact_str/blob/302c595d41373943dc432be2381b8f6a74f0ddf6/compact_str/src/repr/mod.rs#L445-L453

I believe an empty string is always stored inline. Therefore it could be reduced to:

```rs
pub fn is_empty(&self) -> bool {
  self.last_byte() == LastUtf8Char::L0 as u8
}
```

If my assumption is correct, let me know and I'll make a PR.

**Comment by NobodyXu** [comment]
It's almost correct, except when API [`CompactString::from_string_buffer`](https://docs.rs/compact_str/latest/compact_str/struct.CompactString.html#method.from_string_buffer) is called, it can be used to construct an empty string allocated on heap.

IMO you can fix this by changing `from_string_buffer` to special case empty string, but I'm not sure if @ParkMyCar or @Kijewski agrees on this since this does make it less useful for its intended use case.

**Comment by Kijewski** [comment]
It would fail for [`CompactString::with_capacity()`](https://docs.rs/compact_str/latest/compact_str/struct.CompactString.html#method.with_capacity), too.

**Comment by NobodyXu** [comment]
> It would fail for [`CompactString::with_capacity()`](https://docs.rs/compact_str/latest/compact_str/struct.CompactString.html#method.with_capacity), too.

In that case, maybe we should check for capacity and length, instead of modifying `from_string_buffer`?

**Comment by overlookmotel** [comment]
Oh dear. Yes, I completely failed to consider that capacity and length are independent, so zero-length strings can be on the heap.

In that case, I don't think there's anything to be done here.

If static string didn't have it's own discriminant, then this would work:

```rs
pub fn is_empty(&self) -> bool {
    let heap_len = self.1;
    let last_byte = self.last_byte(); 
    // Empty inline string also contains 0 in the bytes holding heap length
    heap_len == 0 && (last_byte == 192 || last_byte == 224)
}
```

That produces branch-free code which *might* be a bit faster. But adding `|| last_byte == 225` does result in a branch, so it's no good.

Unless anyone else sees any mileage in this, I'll close the issue.

**Comment by Kijewski** [comment]
Yeah, I guess this particular method is as good as it gets for the current data representation.

But nevertheless, thank you for looking into possible optimizations! It is very easy to convince yourself that whatever you implemented is the optimum, so it's always welcome to get a new perspective to challenge your view and maybe teach you something new!

**Comment by NobodyXu** [comment]
> If static string didn't have it's own discriminant, then this would work:

I think you code might still work, since static string is guaranteed to be non-empty now (inlined if empty) and its layout is stable:

https://github.com/ParkMyCar/compact_str/blob/302c595d41373943dc432be2381b8f6a74f0ddf6/compact_str/src/repr/static_str.rs#L20

**Comment by overlookmotel** [comment]
> I think you code might still work, since static string is guaranteed to be non-empty now (inlined if empty) and its layout is stable

Unfortunately I was out of date - the value of `HEAP_MASK` is now 216 not 224. Changing my 2nd version to `last_byte == 192 || last_byte == 216` produces a branch, because testing for "192 or 216" can't be done with an `and` operation the way "192 or 224" can.

I think I'll give up at this point! But thanks both for entertaining my attempts.

**Comment by overlookmotel** [comment]
Oh wait! Maybe I've got something:

```rs
const INLINE_MASK: u8 = 192;
const HEAP_MASK: u8 = 216;
const HEAP_MASK_AFTER_SUB: usize = (HEAP_MASK as usize)
  .wrapping_sub(INLINE_MASK as usize);

pub fn is_empty_new(&self) -> bool {
  let len_heap = ensure_read(self.1);
  let mut len = self.last_byte().wrapping_sub(INLINE_MASK as usize);
  if len >= HEAP_MASK_AFTER_SUB {
    len = len_heap;
  }
  len == 0
}
```

If I'm not mistaken (but very possible I am), this uses 1 less register.

https://godbolt.org/z/Y1v79518x

If this *is* more optimal, could maybe also be used by `len()` and `as_slice()`:

```rs
const STATIC_STR_MASK_AFTER_SUB: usize = STATIC_STR_MASK as usize
  .wrapping_sub(LENGTH_MASK as usize);

let mut length = core::cmp::min(
  self.last_byte().wrapping_sub(LENGTH_MASK) as usize,
  STATIC_STR_MASK_AFTER_SUB,
);
// `length` now is either inline length or 24/25 for heap/static string
let heap_length = self.1;
if length > MAX_SIZE {
  length = heap_length;
}
```

or, if `&` is cheaper than `cmp::min`:

```rs
let mut length = (self.last_byte().wrapping_sub(LENGTH_MASK) as usize)
  & 31;
let heap_length = self.1;
if length > MAX_SIZE {
  length = heap_length;
}
```

**Comment by overlookmotel** [comment]
Or is this better?

```rs
const LENGTH_MASK: u8 = 192;
const HEAP_MASK: u8 = 216;
const HEAP_MASK_AFTER_SUB: usize = (HEAP_MASK as usize)
  .wrapping_sub(LENGTH_MASK as usize);

pub fn is_empty_new2(&self) -> bool {
  let len_heap = self.1;
  let last_byte = self.last_byte() as usize;
  let len = last_byte.wrapping_sub(LENGTH_MASK as usize);
  let is_empty_inline = len == 0;
  let is_empty_heap = len == HEAP_MASK_AFTER_SUB && len_heap == 0;
  is_empty_inline || is_empty_heap
}
```

Same instruction count, but `ensure_read(self.1)` is gone, and there's no branches.

https://godbolt.org/z/T1xPT3WzG
